catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/builder/XmlModelBuilder.java

```diff
@@ -109,87 +75,194 @@ public JAXBElement getResult() {
+            supplierInProgress);
+    return this;
+  }
+
+  @Override
+  public XmlModelBuilder property(String name) {
+    setProperty(name);
+    return this;
+  }
+
+  @Override
+  public XmlModelBuilder value(String value) {
+    setValue(value);
     return this;
   }
 
   @Override
-  public FlatFilterBuilder beginBinaryTemporalType(String operator) {
+  public XmlModelBuilder value(boolean value) {
+    // TODO - How can we properly marshal Boolean values without resorting to this?
+    setValue(Boolean.toString(value));
+    return this;
+  }
+
+  private void beginBinaryLogicType(NodeReducer<JAXBElement<?>> reducer) {
     verifyResultNotYetRetrieved();
     verifyTerminalNodeNotInProgress();
-    MultiNodeReducer temporalMapping = TERMINAL_OPS.get(operator);
-    if (temporalMapping == null) {
-      throw new IllegalArgumentException(
-          "Cannot find mapping for binary temporal operator: " + operator);
-    }
-    supplierInProgress = new TerminalNodeSupplier(temporalMapping);
-    return null;
+    logicOpCache.push(reducer);
+    depth.push(new ArrayList<>());
   }
 
-  @Override
-  public XmlModelBuilder beginBinarySpatialType(String operator) {
+  private void beginTerminalType(NodeReducer<JAXBElement<?>> reducer) {
+    verifyResultNotYetRetrieved();
+    verifyTerminalNodeNotInProgress();
+    supplierInProgress = new PropertyValueNodeSupplier<>(reducer);
+  }
+
+  /**
+   * Performs necessary validation and state transition to finish a binary logic type's construction
+   * that is in progress.
+   */
+  private void endBinaryLogicType() {
     verifyResultNotYetRetrieved();
     verifyTerminalNodeNotInProgress();
-    MultiNodeReducer spatialMapping = TERMINAL_OPS.get(operator);
-    if (spatialMapping == null) {
-      throw new IllegalArgumentException(
-          "Cannot find mapping for binary spatial operator: " + operator);
+    verifyLogicalNodeInProgress();
+    verifyLogicalNodeHasChildren();
+    verifyLogicalNodeHasEnoughChildrenPerSchema();
+    JAXBElement result = logicOpCache.pop().apply(depth.pop());
+    if (depth.isEmpty()) {
+      rootNode = result;
+    } else {
+      depth.peek().add(result);
     }
-    supplierInProgress = new TerminalNodeSupplier(spatialMapping);
-    return this;
   }
 
-  @Override
-  public XmlModelBuilder endTerminalType() {
+  /**
+   * Performs necessary validation and state transition to finish a function type's construction
+   * that is in progress, particularly for support of nested functions.
+   */
+  private void endFunctionType() {
+    verifyResultNotYetRetrieved();
+    verifyTerminalNodeInProgress();
+    NodeSupplier<JAXBElement<?>> parent = supplierInProgress.getParent();
+    if (parent == null) {
+      throw new IllegalStateException(
+          "Null parent should not have passed the check in the end() method, "
+              + "verify the implementation of XmlModelBuilder for errors");
```

Comment:
✏️ This might make more sense as an `IllegalArgumentException` since the check occurs elsewhere and has separate semantics associated with it.